### ISSUE TITLE
Error when running in parallel

### ISSUE THREAD

**Issue by araikes** [BIDS-Apps/MRtrix3_connectome#54]
Simple question I hope. Is there a way to pick up where something left off if stuff fails? For instance, I was running a subject in parallel with a lot of other people (that maxed out our 80 core system and I'm hoping that that's the issue) and I got an error of `std::bad_alloc` on this step:

```
mrcalc 1 dwi_meanbzero.mif -div dwi_mask.mif -mult - | mrhistmatch nonlinear - T1_biascorr_brain.mif dwi_pseudoT1.mif -mask_input dwi_mask.mif -mask_target T1_mask.mif
```

I kept all of the intermediate files (verbosity 3 + debug). Is there a way to pick this person up using the temp folders so that I don't need to run topup/eddy again?

**Comment by Lestropie** [comment]
In *MRtrix3*, yes; in the container environment, not easily.

(I suspect) Unlike other BIDS Apps, I've explicitly designed this one to be able to be used either in a container environment, or as a stand-alone *MRtrix3* external project. The script interface changes a bit depending on the environment in which it is run. In particular, when running in a container I remove a lot of the standard *MRtrix3* command-line options that are not so relevant, in order to make the interface as simple as possible and consistent with other BIDS Apps; conversely, when run outside of a container, the script interface is identical to other *MRtrix3* commands.

It's not very well documented, but all *MRtrix3* Python scripts have a command-line option "`-continue`" (see e.g. for [`dwipreproc`](https://mrtrix.readthedocs.io/en/latest/reference/scripts/dwipreproc.html#standard-options)), which is intended for precisely this kind of situation: essentially resuming the operation of a script from some point of failure. You provide the location of the script scratch directory, and the name of the last file that was successfully generated, and it simply skips all processing steps up to and including that point. It's clumsy, and I'd [prefer something better](https://github.com/MRtrix3/mrtrix3/issues/713), but it's been useful enough for me to have put the effort into supporting it.

This is however one of the command-line options that I explicitly remove when the script runs in a container environment, on the prediction that it would be more likely to cause confusion than it would be to actually be useful to anyone operating in the container environment. Maybe that prediction is off, but that was the logic at the time.

So if you can run the script as a standalone *MRtrix3* executable (some advice [here](https://mrtrix.readthedocs.io/en/latest/troubleshooting/FAQ.html#making-use-of-python-scripts-library)), you should be able to exploit the `-continue` option. In the container environment you could only do it if you modified the code to prevent deletion of the command-line option that you need (code [here](https://github.com/BIDS-Apps/MRtrix3_connectome/blob/0.4.2/mrtrix3_connectome.py#L1044-L1045)).

**Comment by araikes** [comment]
Alright... I keep playing with it.

As a follow up since at the moment I'll need to rerun the participants that are stalled out, is the script running recon-all twice? See lines 550 and 551.

Thanks for the greater support as always @Lestropie

**Comment by Lestropie** [comment]
> is the script running recon-all twice? See lines 550 and 551.

> ```
> run.command('recon-all -sd ' + app.tempDir + ' -subjid freesurfer -i T1_registered.nii')
> run.command('recon-all -sd ' + app.tempDir + ' -subjid freesurfer -all' + reconall_multithread_options)
> ```

Nope, that's standard `recon-all` usage: the first generates a new subject and imports the T1 image, the second runs all reconstruction steps. Don't know why they chose to have a single command do both :man_shrugging:

**Comment by Lestropie** [comment]
Closing since a `std::bad_alloc` could occur for any of a number of reasons; but re-raise or open a new issue if a similar problem is  encountered in `0.5.0`.
